### ISSUE TITLE
Can't load file on device

### ISSUE THREAD

**Issue by maurovisintin** [zmxv/react-native-sound#89]
Hi,
I'm working on a simple soundboard app (source code is available at https://github.com/sexykodo/mlcSoundboard) that  automatically fetches remote audio files according to a JSON and puts them in MainBundle with RNFetchBlob.

It's something like this:

```
fetchRemoteFile = (obj) => {
        RNFetchBlob
            .config({
                path : RNFetchBlob.fs.dirs.MainBundleDir + encodeURI(obj.Path) //target path
            })
            .fetch('GET', MAIN_URL + encodeURI(obj.Path), {})
            .then((res) => {
                // the conversion is done in native code
                // the following conversions are done in js, it's SYNC
                console.log('The file saved to ', res.path())
            })
            // Status code is not 200
            .catch((errorMessage, statusCode) => {
                // error handling
                console.log(errorMessage)
            })
    }

```

In this way I can dynamically sync local files with remote ones (I want the soundboard to work offline).
Everything works like a charm on IOS Simulator.

On physical device though, when I load the local file I get an errorcode "ENSOSSTATUSERRORDOMAIN2003334207",
wich is similar to #64 and probably caused by files not being included in Resources.

Is it actually possible to programmatically save an audio file (with RNFetchBlob) and play it?

Thanks in advance

**Comment by GroupXTech** [comment]
We have a similar project, was working until RN 0.39. Have you figured this out?

**Comment by sexykodo** [comment]
Ye, I got this working with  RNFetchBlob.fs.dirs.DocumentDir

**Comment by nishiltamboli** [comment]
I got it working in Android with a fully resolved absolute path. However the same doesnt work in iOS. Instead, the mentioned implementation with only the filename and Path constant works in iOS. I am storing files in Documents Directory.

So I ended up writing a Platform specific code for path.

- Android - Fully resolved Absolute Path to the file
- iOS - Only the file name

For instance, local_link = (Platform.OS === 'ios') ? filename : downloadDest;
So, downloadDest is something like "/data/user/0/com.app/files/AUD0001.mp3"
And filename is like "AUD0001.mp3"

Anybody has any idea why is there such an issue?

**Comment by cooperka** [comment]
I was getting the same error when trying to play the file right away. This workaround fixes it for me (but it's hacky):

```js
setTimeout(() => {
  var sound = new Sound(this.state.audioPath, '', (error) => {
    if (error) { /* ... */ }
  });

  setTimeout(() => {
    sound.play((success) => {
      if (success) { /* ... */ }
      else { /* ... */ }
    });
  }, 100);
}, 100);
```

**Comment by dturton** [comment]
Still not working for me. Now I don't get any console logs, probably because of the timeouts. Here is my code, does it look right?

`        setTimeout(() => {
          var sound = new Sound(this.state.audioPath, '', (error) => {
            if (error) {
              console.log('failed to load the sound', error);
            }
          });

          setTimeout(() => {
            sound.play((success) => {
              if (success) {
                 console.log('successfully finished playing');
               } else {
                 console.log('playback failed due to audio decoding errors');
               }
            });
        }, 100);
      }, 500);`

**Comment by cooperka** [comment]
The timeouts shouldn't affect the console logs, that's really strange. Code looks right though.

**Comment by jimthedev** [comment]
@cooperka did this work for you on a device or in the simulator. Seems to work in the simulator but on an iOS 7 device .play never gets triggered. :(

**Comment by cooperka** [comment]
@jimthedev: Working for me on both device and simulator. There was also another issue that could be related for you, fixed here: https://github.com/jsierles/react-native-audio/pull/155.

**Comment by jimthedev** [comment]
Woot I'll check it out.

**Comment by jimthedev** [comment]
@cooperka Looks like https://github.com/jsierles/react-native-audio/pull/155 solved my issue. Thanks.

**Comment by tonylinyy** [comment]
yes, jsierles/react-native-audio#155 solved my issue

**Comment by dmonay** [comment]
@zmxv @cooperka I'm currently implementing the `setTimeout` workarounds. I have v3.2.2 of `react-native-audio`, released after the PR that seems to fix it for others https://github.com/jsierles/react-native-audio/pull/155. 

Is there something I should be doing on my end to fix it? Is that why this issue is still open?

**Comment by wandonye** [comment]
Confused by the discussion above. So you guys switched to `react-native-audio` ?

**Comment by dmonay** [comment]
@wandonye I think most of us are using the two libraries in conjunction.

**Comment by fredbt** [comment]
Still suffering with this to play remote files in both the Android emulator and the real devices. It does not work without the timeouts.

**Comment by davidroeca** [comment]
I've come across this same issue in iOS simulator, resolved with a 10ms timeout (might have to increase for slower devices). Any thoughts on wrapping this with a promise?

**Comment by trepidity** [comment]
take a look at #334

**Comment by MehdiJarraya** [comment]
@maurovisintin Hi Maurovisintin, can you provide please example how can I save local sound and read it with RNFetchBlob and react-native-sound.

**Comment by maurovisintin** [comment]
Something like this
`const MAIN_DIR = RNFetchBlob.fs.dirs.DocumentDir`
`const MAIN_URL = 'someurl'`
`fetchRemoteFile = (obj) => {
        return RNFetchBlob
            .config({
                path: `${MAIN_DIR}/${obj.Path}` //local target path
            })
            .fetch('GET', `${MAIN_URL}/${obj.Path}`, {})
            .then(res => {
                console.log('The file saved to ', res.path())
            })
            // Status code is not 200
            .catch((errorMessage, statusCode) => {
                // error handling
            })
    }`

`playSound  = (file) => {
        const sound = new Sound(file.Path, MAIN_DIR, error => {
            if (error) {
              console.log('failed to load the sound', error);
            } else { // loaded successfully
              console.log('duration in seconds: ' + sound.getDuration() +
                  'number of channels: ' + sound.getNumberOfChannels());
                    
                  console.log("name", sound._filename)
                  sound.play(success => {
                      if (success) {
                          console.log("success")
                      } else {
                          console.log(err)
                      }
                  },
                  err => {
                      console.log(err)
                  })
            }
        })

    }`

**Comment by MehdiJarraya** [comment]
@maurovisintin, thank you for your response.
Your solution didn't work for me but I figured out the problem and got it resolved this way:
RNFetchBlob.config({
        fileCache: true,
      }).....
 when initializing the sound
 const sound = new Sound(file.Path,'', error => {....
The sound bundle must be an empty string '' in iOS.

**Comment by mklb** [comment]
maybe this helps: https://github.com/zmxv/react-native-sound/issues/178#issuecomment-427395055

**Comment by tuanduongdn1504** [comment]
> Something like this
> 
> ```
> const MAIN_DIR = RNFetchBlob.fs.dirs.DocumentDir
> const MAIN_URL = 'someurl'
> fetchRemoteFile = (obj) => {
>         return RNFetchBlob
>             .config({
>                 path: `${MAIN_DIR}/${obj.Path}` //local target path
>             })
>             .fetch('GET', `${MAIN_URL}/${obj.Path}`, {})
>             .then(res => {
>                 console.log('The file saved to ', res.path())
>             })
>             // Status code is not 200
>             .catch((errorMessage, statusCode) => {
>                 // error handling
>             })
>     }
> 
> playSound  = (file) => {
>         const sound = new Sound(file.Path, MAIN_DIR, error => {
>             if (error) {
>               console.log('failed to load the sound', error);
>             } else { // loaded successfully
>               console.log('duration in seconds: ' + sound.getDuration() +
>                   'number of channels: ' + sound.getNumberOfChannels());
>                     
>                   console.log("name", sound._filename)
>                   sound.play(success => {
>                       if (success) {
>                           console.log("success")
>                       } else {
>                           console.log(err)
>                       }
>                   },
>                   err => {
>                       console.log(err)
>                   })
>             }
>         })
> 
>     }
> ```

This way resolved issue for both IOS and Android

**Comment by paulmelnikow** [comment]
If there's a reproducible recipe here, it would be useful if someone could open a pull request to add this to the documentation!

**Comment by richardkyeung** [comment]
Just adding one more hacky solution to this thread :) 

cooperka's solution didn't work for me. So I took the ultimate hacky path and just re-fetched the audio file on error.

**Comment by ivanfranchi** [comment]
As August 2019, I got it working with  

`whoosh = new Sound('pre_alarm.mp3', Sound.MAIN_BUNDLE, (error) => {  
        if (error) {  
            console.log('failed to load the sound', error);  
            alert("iOS 1? " + JSON.stringify(error));  
        }  
    });`  

and  

`whoosh.play((success) => {  
                if (success) {  
                    console.log('successfully finished playing');  
                } else {  
                    console.log('playback failed due to audio decoding errors');  
                    alert("iOlayback failed due to audio decoding errors");  
                }  
            });`  

Basically the full name of the sound, Sound.MAIN_BUNDLE and placing the mp3 file in the root directory using the xcode way "add files to project". No timeouts

**Comment by kkusanagi** [comment]
Use 'new Promise' to solve the flow problem

```
async function playsound (file) {
      const sound = null;
      const error = await new Promise(function (resolve,reject){
          sound = new Sound(file.Path, '',resolve);
     });
           if (error) {
              console.log('failed to load the sound', error);
            } else { // loaded successfully
              console.log('duration in seconds: ' + sound.getDuration() + 'number of channels: ' + sound.getNumberOfChannels());
                  console.log("name", sound._filename)
                  sound.play(success => {
                      if (success) {
                          console.log("success")
                      } else {
                          console.log(err)
                      }
                  },
                  err => {
                      console.log(err)
                  });
}
```

**Comment by creativemind1** [comment]
_duration getting in negative with local file but working perfectly fine with server files.[Only for Android].

**Comment by zscafeman** [comment]
execuse me, can provide some sample code for play remote vedio?

**Comment by mynameisliuyurong** [comment]
1.检查RNSound.xcodeproj是否导入
2.检查libRNSOund.a是否添加
3.检查音频文件是否导入
4.初始化看起来像这样：
const mainBundle = Platform.OS === 'ios'
    ? encodeURIComponent(Sound.MAIN_BUNDLE)
    : Sound.MAIN_BUNDLE;
        new Sound('music.mp3', mainBundle, (error) => { ... }
5.祝您成功！

**Comment by AkilUnik** [comment]
import ReactNativeBlobUtil from 'react-native-blob-util';
import {Platform, PermissionsAndroid} from 'react-native';

/// grant permission in android

export const downloadFile = async (url, headers) => {
  // Get the app's cache directory
  const {fs} = ReactNativeBlobUtil;
  const cacheDir = fs.dirs.DownloadDir;

  // Generate a unique filename for the downloaded image
  const filename =
    url.split('/').pop().slice(0, 7) + Math.floor(Math.random() + Date.now());
  const imagePath = `${cacheDir}/${filename}`;

  try {
    // Download the file and save it to the cache directory
    const configOptions = Platform.select({
      ios: {
        fileCache: true,
        path: imagePath + '.mp3',
        appendExt: filename.split('.').pop(),
      },
      android: {
        fileCache: true,
        appendExt: filename.split('.').pop(),
        addAndroidDownloads: {
          // Related to the Android only
          useDownloadManager: true,
          notification: true,
          path: imagePath + '.mp3',
          description: 'File',
        },
      },
    });

    const response = await ReactNativeBlobUtil.config(configOptions).fetch(
      'GET',
      url,
      headers,
    );

    // Return the path to the downloaded file
    return response;
  } catch (error) {
    console.error(error);
    return null;
  }
};
and my app.js looks like 

const AudioPlayer = () => {
  const [audioUrl, setAudioUrl] = useState(null);
  const [loading, setLoading] = useState(false);

  const playSound = () => {
    if (audioUrl) {
      let whoosh = new Sound(audioUrl, '', error => {
        if (error) {
          console.log('failed to load the sound', error);
          return;
        }
        whoosh.play(success => {
          if (success) {
            console.log('successfully finished playing');
          } else {
            console.log('playback failed due to audio decoding errors');
          }
        });
      });
    }
  };
  useEffect(() => {
    Sound.setCategory('Playback');
  }, []);

  const fetchAudioFile = async () => {
    try {
      setLoading(true);
      if (Platform.OS === 'android') {
        console.log('android');
        downloadFile(BaseUrl, {
          Authorization: Token,
        }).then(res => {
          const audioUrlfilePath = 'file://' + res.path();
          console.log('Downlaod file path ===>', res.path());
          setAudioUrl(audioUrlfilePath);
        });
      } else {
        downloadFile(BaseUrl, {
          Authorization: Token,
        }).then(res => {
          ReactNativeBlobUtil.ios.previewDocument(res.path());
          const audioUrlfilePath = 'file://' + res.path();
          console.log('Downlaod file path ===>', res.path());
          setAudioUrl(audioUrlfilePath);
        });
      }
    } catch (e) {
      console.log(e);
    } finally {
      setLoading(false);
    }
  };

  return (
    <View style={{flex: 1, justifyContent: 'center', alignItems: 'center'}}>
      <TouchableOpacity
        onPress={fetchAudioFile}
        style={{
          borderWidth: 1,
          paddingHorizontal: RFValue(17),
          borderRadius: RFValue(10),
          paddingVertical: RFValue(10),
        }}>
        {loading ? <Text>Downloading....</Text> : <Text>Download audio </Text>}
      </TouchableOpacity>
      <TouchableOpacity
        disabled={audioUrl === null}
        style={{
          borderWidth: 0,
          paddingVertical: RFValue(15),
          paddingHorizontal: RFValue(25),
          width: RFValue(100),
          borderRadius: RFValue(10),
          marginTop: RFValue(10),
          backgroundColor: 'red',
          justifyContent: 'center',
          alignItems: 'center',
          flexDirection: 'row',
          gap: RFValue(20),
          opacity: audioUrl === null ? 0.5 : 1,
        }}
        onPress={playSound}>
        <Image
          source={playIcon}
          style={{
            height: RFValue(20),
            width: RFValue(20),
            resizeMode: 'contain',
          }}
        />
      </TouchableOpacity>
    </View>
  );
};

**Comment by AkilUnik** [comment]
```const AudioPlayer = () => {
  const [audioUrl, setAudioUrl] = useState(null);
  const [loading, setLoading] = useState(false);

  const playSound = () => {
    if (audioUrl) {
      let whoosh = new Sound(audioUrl, '', error => {
        if (error) {
          console.log('failed to load the sound', error);
          return;
        }
        whoosh.play(success => {
          if (success) {
            console.log('successfully finished playing');
          } else {
            console.log('playback failed due to audio decoding errors');
          }
        });
      });
    }
  };
  useEffect(() => {
    Sound.setCategory('Playback');
  }, []);

  const fetchAudioFile = async () => {
    try {
      setLoading(true);
      if (Platform.OS === 'android') {
        console.log('android');
        downloadFile(BaseUrl, {
          Authorization: Token,
        }).then(res => {
          const audioUrlfilePath = 'file://' + res.path();
          console.log('Downlaod file path ===>', res.path());
          setAudioUrl(audioUrlfilePath);
        });
      } else {
        downloadFile(BaseUrl, {
          Authorization: Token,
        }).then(res => {
          ReactNativeBlobUtil.ios.previewDocument(res.path());
          const audioUrlfilePath = 'file://' + res.path();
          console.log('Downlaod file path ===>', res.path());
          setAudioUrl(audioUrlfilePath);
        });
      }
    } catch (e) {
      console.log(e);
    } finally {
      setLoading(false);
    }
  };

  return (
    <View style={{flex: 1, justifyContent: 'center', alignItems: 'center'}}>
      <TouchableOpacity
        onPress={fetchAudioFile}
        style={{
          borderWidth: 1,
          paddingHorizontal: RFValue(17),
          borderRadius: RFValue(10),
          paddingVertical: RFValue(10),
        }}>
        {loading ? <Text>Downloading....</Text> : <Text>Download audio </Text>}
      </TouchableOpacity>
      <TouchableOpacity
        disabled={audioUrl === null}
        style={{
          borderWidth: 0,
          paddingVertical: RFValue(15),
          paddingHorizontal: RFValue(25),
          width: RFValue(100),
          borderRadius: RFValue(10),
          marginTop: RFValue(10),
          backgroundColor: 'red',
          justifyContent: 'center',
          alignItems: 'center',
          flexDirection: 'row',
          gap: RFValue(20),
          opacity: audioUrl === null ? 0.5 : 1,
        }}
        onPress={playSound}>
        <Image
          source={playIcon}
          style={{
            height: RFValue(20),
            width: RFValue(20),
            resizeMode: 'contain',
          }}
        />
      </TouchableOpacity>
    </View>
  );
};
